src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+                    {
+                        _projectReferencesAddedInBatch.Add(projectReference);
+                    }
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(w => w.OnProjectReferenceAdded(Id, projectReference));
+                }
+            }
+        }
+
+        public bool ContainsProjectReference(ProjectReference projectReference)
+        {
+            if (projectReference == null)
+            {
+                throw new ArgumentNullException(nameof(projectReference));
+            }
+
+            lock (_gate)
+            {
+                if (_projectReferencesRemovedInBatch.Contains(projectReference))
+                {
+                    return false;
+                }
+
+                if (_projectReferencesAddedInBatch.Contains(projectReference))
+                {
+                    return true;
+                }
+
+                return _workspace.CurrentSolution.GetProject(Id).AllProjectReferences.Contains(projectReference);
+            }
+        }
+
+        public IReadOnlyList<ProjectReference> GetProjectReferences()
+        {
+            lock (_gate)
+            {
+                // If we're not batching, then this is cheap: just fetch from the workspace and we're done
+                var projectReferencesInWorkspace = _workspace.CurrentSolution.GetProject(Id).AllProjectReferences;
+
+                if (_activeBatchScopes == 0)
+                {
+                    return projectReferencesInWorkspace;
+                }
+
+                // Not, so we get to compute a new list instead
+                var newList = projectReferencesInWorkspace.ToList();
+                newList.AddRange(_projectReferencesAddedInBatch);
+                newList.RemoveAll(p => _projectReferencesRemovedInBatch.Contains(p));
+
+                return newList;
+            }
+        }
+
+        public void RemoveProjectReference(ProjectReference projectReference)
+        {
+            if (projectReference == null)
+            {
+                throw new ArgumentNullException(nameof(projectReference));
+            }
+
+            lock (_gate)
+            {
+                if (_activeBatchScopes > 0)
+                {
+                    if (!_projectReferencesAddedInBatch.Remove(projectReference))
+                    {
+                        _projectReferencesRemovedInBatch.Add(projectReference);
+                    }
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(w => w.OnProjectReferenceRemoved(Id, projectReference));
+                }
+            }
+        }
+
+        #endregion
+
+        public void RemoveFromWorkspace()
+        {
+            _documentFileChangeContext.Dispose();
+
+            // TODO: clean up open files
+            _workspace.ApplyChangeToWorkspace(w => w.OnProjectRemoved(Id));
```

Comment:
>ApplyChangeToWorkspace [](start = 23, length = 22)

Is ApplyChangeToWorkspace needed here? OnProjectRemoved already takes the workspace lock.